README.md
# LEARNING을 통한 태풍경로 예측

---

최근 태풍들의 위치정보들을 통해 3시간 후 태풍의 위치를 예측하는 머신러닝을 사용합니다.



### FACTORS

---

함수를 모델링하기 위해서 저희는 태풍의 위치(latitude, longitude)와, 그 시간대의 수온(WT)과 풍향(WD)과 풍속(WV)을 이용했습니다. 공공기관에서 제공하는 데이터 특성상 시간대의 간격은 3시간 입니다. 



※ WT는 섭씨온도

※ WD는 0º ~ 360º 의 값을 가지는 극좌표계

※ WV는 km/h 단위





### LEARNING

---

향후 추가 바람.



## 파일들 목록

---

1. AccessDB.py
   
   - 원하시는 서버 mysql에 저장된 태풍들의 정보들을 보관할만한 sql문을 작성해줍니다.
   
2. InputDB.py
   
   * 원하시는 서버의 mysql 내부 database에 최근 태풍들의 위치정보를 input 합니다.
   
3. TyphoonList.py
   
   * 2001년부터 2019년 까지의 우리나라의 영향태풍들의 index와 이름을 크롤링합니다.
   
4. TyphoonListfunc.py
   
   * TyphoonList.py에서 받아온 태풍의 이름, index를 통해 태풍의 위치정보를 크롤링합니다.
   
5. CrawlingInputValues.py

   * [https://earth.nullschool.net](https://earth.nullschool.net/)에서 풍향, 풍속, 수온 데이터를 크롤링하여

     현재 Typhoon 정보가 담겨있는 DB에 UPDATE 합니다.

ChangeLog.md
# Change Log

**Remain Task:**

- 풍향(WD), 수온(WT)에 대한 Factor 정보를 수집해야합니다.
- 정확한 알고리즘을 DB내 태풍 data에 적용시켜야합니다.
- 모델링한 함수를 통해 예시를 시각화 해야합니다.



1.0.0

---

* DB에 태풍 데이터베이스를 만들고 최근 태풍 위치값을 INSERT하는 sql문을 실행하는 py문 작성.

* 최근 태풍 위치 자료들을 기상자료센터에서 크롤링하는 py문 작성.





1.0.1

---

* 2001년 부터 2019년의 영향태풍을 2014년 부터 2019년의 영향태풍으로 변경.

* [https://earth.nullschool.net](https://earth.nullschool.net/)에 나와있는 지난 5년간의 대기상황을 크롤링하여

  5년간 일어난 태풍 경로와 시간에 풍향, 풍속, 온도 값을 삽입하는 py문 작성.



InputDB.py
# -*- coding: utf-8 -*-
"""
Created on Tue Apr  9 19:56:07 2019

@author: sangwon
"""

import pymysql
from TyphoonList import *
from TyphoonListfunc import *

# db_connect information
_host = '168.131.39.95'
_port = 3306
_user = 'root'
_passwd = 'jslab0963'
_db = 'TYP'

# access
db = pymysql.connect(host=_host, port=_port, user=_user, passwd=_passwd, db=_db, charset='utf8')

# 커서 가져오기
cursor = db.cursor()

result = callTyphoonInformation(index)
count = 0

for rst in result:
    
    result2 = rst['information']

    for idx in result2:
        
        sql = "INSERT INTO " + name_index[count] + " (Date, Latitude, Longitude) VALUES (%s, %s, %s);"  
          
        # 실행하기
        cursor.execute(sql, (idx['Date'], idx['latitude'], idx['longitude']))
        
        # DB에 Complete 하기
        db.commit()

    count += 1

# DB 연결 닫기
db.close()

AccessDB.py
# -*- coding: utf-8 -*-
"""
Created on Tue Apr  9 16:50:12 2019

@author: sangwon
"""

import pymysql
from TyphoonList import *

# db_connect information
_host = '168.131.39.95'
_port = 3306
_user = 'root'
_passwd = 'jslab0963'
_db = 'TYP'

# access
db = pymysql.connect(host=_host, port=_port, user=_user, passwd=_passwd, db=_db, charset='utf8')

# 커서 가져오기
cursor = db.cursor()

for idx in name_index:
    typhoonname = str(idx);
    sql = " CREATE TABLE " + typhoonname + " (\
            Date DATETIME NOT NULL,\
            Latitude DECIMAL(4, 1) NOT NULL,\
            Longitude DECIMAL(4, 1) NOT NULL,\
            WD INT UNSIGNED,\
            WV DECIMAL(6,3),\
            WT DECIMAL(6,3),\
            PRIMARY KEY(Date)); "
        
    # 실행하기
    cursor.execute(sql)
    
    # DB에 Complete 하기
    db.commit()


# DB 연결 닫기
db.close()






TyphoonList.py
#!/usr/bin/env python
# coding: utf-8
import requests
from bs4 import BeautifulSoup

index = [201825, 201824, 201819, 201818, 201807, 201718, 201705,201703, 201618,\
         201616, 201515, 201512, 201511, 201511, 201509, 201419, 201412, 201411, 201408]
'''
factor Crawling 하는데 2013년 기록은 존재하지 않아서 삭제

          201324, 201315, 201304, 201216, 201215, 201214, 201210, 201207, 201112,\
          201109, 201105, 201009, 201007, 201004, 200807, 200711, 200705, 200704,\
          200613, 200610, 200603, 200514, 200418, 200416, 200415, 200410, 200407,\
          200314, 200310, 200306, 200304, 200215, 200209, 200208, 200205, 200111]
'''


def getTyphoonName(index): 
    typhoon_name = []
    for idx in index:
        res = requests.get('http://typ.kma.go.kr/TYPHOON/ko/weather/typhoon_06_pop.jsp?' +\
                           'json={%22typYear%22:' +\
                           str(int(idx / 100)) +\
                           ',%22typSeq%22:' +\
                           str(idx % 100) + '}')
        
        soup = BeautifulSoup(res.content, 'html.parser')    
        title = soup.select('div > h1.searchH1')
        fullname = title[0].get_text().strip()
        namesplit = []
        namesplit = fullname.split('(')[1].split(')')[0]
        typhoon_name.append(namesplit)
        #typhoon_name.append(name)
    
    return typhoon_name

name_index = []
beforeNameIndex = getTyphoonName(index)

for idx in beforeNameIndex:
    name_index.append(idx.replace("-",""))    


name_index[0] = 'KONGREY2'
name_index[8] = 'CHABA2'
name_index[15] = 'NAKRI2'



TyphoonListfunc.py
# coding: utf-8
import requests
from bs4 import BeautifulSoup
from TyphoonList import *
from pprint import pprint


def getTyphoonInformation(index): 
    typhoon_dic = {}
    typhoon_arr = []
    res = requests.get('http://typ.kma.go.kr/TYPHOON/ko/weather/typhoon_06_pop.jsp?' +\
                       'json={%22typYear%22:' +\
                       str(int(index / 100)) +\
                       ',%22typSeq%22:' +\
                       str(index % 100) + '}')
    
    soup = BeautifulSoup(res.content, 'html.parser')    
    title = soup.select('div > h1.searchH1')
    typhoon_dic['name'] = title[0].get_text().strip()
    contents = soup.select('div.table_area > table > tbody > tr > td')
    
    num = 0    
    for cnt in contents:
        num = num + 1
        if (num % 11 == 1) :
            typhoon_inf = {}
            typhoon_inf['Date'] = cnt.get_text().strip()
        elif (num % 11 == 2) : 
            typhoon_inf['latitude'] = cnt.get_text().strip()
        elif (num % 11 == 3) :
            typhoon_inf['longitude'] = cnt.get_text().strip()
            typhoon_arr.append(typhoon_inf)
        
        
    typhoon_dic['information'] = typhoon_arr
    
    return typhoon_dic

def callTyphoonInformation(index_arr):
    typhoons_list = []
    
    for idx in index_arr:
        typhoons_list.append(getTyphoonInformation(idx))

    return typhoons_list



CrawlingInputValues.py
# -*- coding: utf-8 -*-
from bs4 import BeautifulSoup
from selenium import webdriver
from TyphoonList import *
from TyphoonListfunc import *
import time
import pymysql

# db_connect information
_host = '168.131.39.95'
_port = 3306
_user = 'root'
_passwd = 'jslab0963'
_db = 'TYP'

# access
db = pymysql.connect(host=_host, port=_port, user=_user, passwd=_passwd, db=_db, charset='utf8')
cursor = db.cursor()

for ni in name_index:

    sql = "SELECT * FROM " + ni
    cursor.execute(sql)
    result = cursor.fetchall()
    
    for row_data in result:
        _date = str(row_data[0])[0:10].replace("-","/")
        _time = str(row_data[0])[11:16].replace(":","")
        latitude = str(row_data[1])
        longitude = str(row_data[2])
        
        driver = webdriver.Chrome('/Users/sangwon/Documents/chromedriver')
        driver.get('https://earth.nullschool.net'\
                   '/#' + _date + '/' + _time + 'Z'\
                   '/wind/surface/level/overlay=temp/orthographic'\
                   '/loc=' + longitude + ',' + latitude)
        time.sleep(5)        
        html = driver.page_source
                 
        soup = BeautifulSoup(html, 'html.parser')
        WD = (soup.select('#location-wind')[0].text).split(' @ ')[0].replace("°","")
        WV = (soup.select('#location-wind')[0].text).split(' @ ')[1]
        WT = soup.select('#location-value')[0].text
        
        
        sql2 = "UPDATE " + ni + " SET WD = '" + WD + "', WV = '" + WV + "', WT = '" + WT + "' WHERE Date = '" + str(row_data[0]) + "';"
              
        # 실행하기
        cursor.execute(sql2)
        
        # DB에 Complete 하기
        db.commit()
        
        driver.quit()

